contracts/SpokePool.sol

```diff
@@ -39,21 +43,57 @@ abstract contract SpokePool is Testable, Lockable, MultiCaller {
     // Origin token to destination token routings can be turned on or off.
     mapping(address => mapping(uint256 => bool)) public enabledDepositRoutes;
 
+    struct RelayData {
```

Comment:
This is really interesting and raises a tradeoff.

In v1, we store a hash of the full relay data and require the caller to pass it in on each call. In v2, it looks like the current design is to store the relay data. I'm guessing this is because a) it's simpler and b) storage is cheap on L2s and calldata is expensive.

However, this amount of storage might make our L2 -> L1 transfers more expensive. Also, I think the common path through this code is that there will be a single relayer who fills the entire thing on one go, so there will be usually be no need to call into it multiple times to read this data.

Another, related thought is that since calldata is expensive on L2, it will be an added cost to have to multicall initiate and fill relay on a typical relay. This requires more calldata than just having a combined method that can do both things.

Going back to these structs, the _only_ piece of dynamic data is filledAmount. Everything else is static after initialization I think.

Putting this all together, what do you think of a setup where _all_ the static data (everything in both these structs except for filledAmount) is hashed to look up a single number, which is filledAmount?

In this setup you could have a single method called `fill` that takes all this data + the amount you want to fill. If the relay exists, just adds to the filledAmount. If the relay doesn't exist, it initializes it.

Thoughts? Note: I could totally be missing something, I was just trying to think of ways to keep this cheap on both L1 and L2.

Reply:
I didn't go into nearly as much thought about cost when designing this interface. I was motivated mostly by simplicity. 

I think generating the relay hash as it's stored here is simple enough and I also agree we should key fillAmount by relayHash.